### PR TITLE
Restore checkpoint replica from the shared memory of other nodes.

### PR DESCRIPTION
**Pull request by workingloong**
### What changes were proposed in this pull request?

Restore checkpoint replica from the shared memory of other nodes.

### Why are the changes needed?

To restore the training fastly.

### Does this PR introduce any user-facing change?

Yes. A new argument in the `Checkpointer` is added.

### How was this patch tested?

UT.

### CODE REVIEW

File: dlrover/trainer/tests/torch/checkpoint_backup_test.py [hunk 1]
```diff
@@ -87,12 +87,10 @@ def run_checkpoint_backup(rank, world_size):
...
 
     with mock.patch.object(
-        FullCkptBackupManager, "_get_backup_ranks", return_value=[0, 1]
+        FullCkptReplicaManager, "_get_backup_ranks", return_value=[0, 1]
     ):
-        back_manager = FullCkptBackupManager(
-            local_rank=rank, local_world_size=1
-        )
-    shm_tensor, _ = back_manager._gather_owner_checkpoint(shm_hanlder)
+        back_manager = FullCkptReplicaManager(1)
```
Comment (BalaBalaYi) [inline dlrover/trainer/tests/torch/checkpoint_backup_test.py:92]: lint

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 2]
```diff
@@ -335,6 +346,25 @@ def get_state_dict_from_memory(self):
...
+        if (
+            byte_tensor is not None
+            and meta
+            and not self._shm_handler.shared_memory
+        ):
+            shm_size = byte_tensor.size()[0]
+            self._shm_handler.init_shared_memory(create=True, size=shm_size)
+            self._shm_handler.metadata.set(meta)
+            logger.info(
+                "Restore the checkpoint from the replica in the "
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py:363]: Better print the variables for metadata, size, etc.

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 3]
```diff
@@ -335,6 +346,25 @@ def get_state_dict_from_memory(self):
             logger.info(f"Load checkpoint at step {config.step} from memory.")
         return config.step, state_dict
 
+    def _restore_memory_from_replica(self):
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py:349]: Need ut for coverage.

File: dlrover/trainer/torch/flash_checkpoint/replica.py [hunk 4]
```diff
@@ -249,52 +273,77 @@ def backup(self, shm_handler: SharedMemoryHandler):
...
+            group_size, dtype=torch.long, device=self.current_device
+        )
+        shm_size_list = [
+            shm_sizes_tensor[i].unsqueeze(dim=0) for i in range(group_size)
+        ]
+        # Allgather tensor sizes
+        dist.all_gather(shm_size_list, local_size, group=self._backup_group)
+
+        max_tensor_size = int(max(shm_size_list))
+        return max_tensor_size
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/replica.py:349]: ```return int(max(shm_size_list))```

File: dlrover/trainer/torch/flash_checkpoint/replica.py [hunk 5]
```diff
@@ -249,52 +273,77 @@ def backup(self, shm_handler: SharedMemoryHandler):
...
         ]
-        dist.all_gather(output_tensors, flag)
-        src_rank = self.backup_ranks[0]
-        for i, flag in enumerate(output_tensors):
-            if flag == 1:
-                src_rank = self.backup_ranks[i]
-                break
+        dist.all_gather(all_flags, flag, group=self._backup_group)
+
+        replica_not_exist = all([t == zero_tensor for t in all_flags])
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/replica.py:304]: need some annotations

File: dlrover/trainer/torch/flash_checkpoint/engine.py (hunk 2 above)
Comment (workingloong) [reply dlrover/trainer/torch/flash_checkpoint/engine.py:363]: > Better print the variables for metadata, size, etc.

The variables may be too many and we can print the size in the log.

**Review by BalaBalaYi — APPROVED** [review]

**Review by samplise — APPROVED** [review]

**Pull request merged**